frame/session/src/historical/storage.rs

```diff
@@ -0,0 +1,138 @@
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Validator Set Extracting an iterator from an offchain worker stored list containing historical validatorsets.
+//!
+//! This is used in conjunction with [`ProvingTrie`](super::ProvingTrie) and
+//! the offchain indexing API.
+
+use sp_io::offchain_index;
+use sp_std::prelude::*;
+
+use super::super::{Module as SessionModule, SessionIndex};
+use super::Trait;
+
+const PREFIX: &[u8] = b"historical";
+const LAST_PRUNE: &[u8] = b"historical_last_prune";
+const HEAD: &[u8] = b"historical_head";
+
+pub struct ValidatorSet<T: Trait> {
+    validator_set: Vec<(T::ValidatorId, T::FullIdentification)>,
+}
+
+/// Derive the key used to store the list of validators
+fn derive_key<P: AsRef<[u8]>, T: Trait>(prefix: P, session_index: Vec<u8>) -> Vec<u8> {
+    assert!(session_index.len() > 0);
+    let prefix = prefix.as_ref();
+    let mut concatenated = Vec::with_capacity(prefix.len() + 1 + session_index.len());
+    concatenated.extend_from_slice(prefix);
+    concatenated.push('/');
+    (&mut concatenated[(prefix.len() + 1)..]).extend_from_slice(session_index.as_slice());
+    concatenated
+}
+
+impl<T: Trait> ValidatorSet<T> {
+    /// Load the set of validators for a paritcular session index from the offchain storage.
+    ///
+    /// If none is found or decodable given `prefix` and `session`, it will return `None`.
+    /// Empty validator sets should only ever exist for genesis blocks.
+    fn load_from_offchain(session_index: SessionIndex) -> Option<Self> {
+        let derived_key = derive_key(STATIC_PREFIX, session_index.encode());
+        let validator_set = StorageValueRef::persistent(derived_key.as_ref())
+            .get::<Vec<(T::ValidatorId, T::FullIdentification)>>()
+            .flatten();
+        validator_set
+    }
+
+    /// Access the underlying `ValidatorId` and `FullIdentification` tuples as slice.
+    fn as_slice(&self) -> &[(T::ValidatorId, T::FullIdentification)] {
+        self.validator_set.as_slice()
+    }
+
+    /// Convert `self` to a vector and consume `self`.
+    fn to_vec(self) -> Vec<(T::ValidatorId, T::FullIdentification)> {
+        self.validator_set
+    }
+
+    /// Prune anything older than the current session index.
+    ///
+    /// For behaviour details refer to [`fn prune_older_than`](Self::prune_older_than).
+    ///
+    /// **Must** be called from the offchain worker.
+    fn prune() {
```

Comment:
Pruning behavior probably wants to lag a certain number of sessions behind the head. Is there support for that? Seems that `prune_older_than` would not need to change

Reply:
You are right, a `delay` parameter makes more sense.